<a href="https://colab.research.google.com/github/MahimaMaurya/Indigo/blob/main/Notebooks/0_Data_Cleansing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install transformers datasets nltk spacy sklearn matplotlib seaborn plotly
!python -m spacy download en_core_web_sm

  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Preparing metadata (setup.py) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 69.8 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


In [2]:
!pip install datasets
import pandas as pd


  Using cached datasets-2.20.0-py3-none-any.whl.metadata (19 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 547.8/547.8 kB 4.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 7.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 316.1/316.1 kB 11.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 39.9/39.9 MB 15.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.9/64.9 kB 2.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 3.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 4.7 MB/s eta 0:00:00
  Attempting uninstall: requests
    Found existing installation: requests 2.31.0
    Uninstalling requests-2.31.0:
      Successfully uninstalled requests-2.31.0
  Attempting uninstall: pyarrow
    Found existing installation: pyarrow 14.0.2
    Uninstalling pyarrow-14.0.2:
      Successfully uninstalled pyarrow-14.0.2
  Attempting uninstall: fsspec
    Found existing insta

In [3]:
from datasets import load_dataset

dataset = load_dataset("toughdata/quora-question-answer-dataset")



/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


Generating train split:   0%|          | 0/56402 [00:00<?, ? examples/s]

In [4]:
# Split the dataset into training and validation sets (80% training, 20% validation)
dataset = dataset['train'].train_test_split(test_size=0.2)

# Check the size of the splits
print(f"Training set size: {len(dataset['train'])}")
print(f"Validation set size: {len(dataset['test'])}")

Training set size: 45121
Validation set size: 11281


In [5]:
dataset['train'][0]
dataset['train'][1]
dataset['train'][2]


{'question': 'If the Von Karman line is 62 miles above, then the center of the Earth should be at minimum 62 miles below, right?',
 'answer': "No. The Karman line has nothing to do with the centre of the Earth.\n From Wikipedia…\n “The FAI defines the Kármán line as space beginning 100 kilometres (54 nautical miles; 62 miles; 330,000 feet) above [LINKED_TEXT: Earth] [URL: https://en.m.wikipedia.org/wiki/Earth]'s [LINKED_TEXT: mean sea level] [URL: https://en.m.wikipedia.org/wiki/Above_mean_sea_level]. This number is roughly based on a theoretical limit to the altitude reachable by a conventional airplane.”\n End quote.\n The atmosphere actually extends much further into space. There is no hard line where the atmosphere ends, it just gets thinner and thinner as you go up. The SOHO spacecraft discovered that the Earth has a hydrogen layer (called the geocorona) which extends well past the Moon.\n"}

In [6]:
dataset['train'].features

{'question': Value(dtype='string', id=None),
 'answer': Value(dtype='string', id=None)}

# Summary of the dataset
The Quora Question Answer Dataset is designed to develop and evaluate question-answering models. The dataset consists of pairs of questions and answers, sourced from Quora, and is suitable for natural language processing tasks such as training models to generate accurate responses to user queries.

Dataset Features:

	•	question: A string representing the question asked.
	•	answer: A string representing the answer to the question.

Dataset Size:

	•	Total entries: 56,402
	•	Training set size: 45,121 (80% of the total data)
	•	Validation set size: 11,281 (20% of the total data)


In [7]:
!pip install pandas openpyxl

In [8]:
import pandas as pd

# Convert the dataset to pandas DataFrame
train_df = pd.DataFrame(dataset['train'])


# Save the DataFrames to Excel files
train_df.to_excel('/content/quora_train.xlsx', index=False)

# Data Assessing
# Issues with the dataset
*   Irrrelavant HTML tags
*   Emojis
*   Presence of \n








In [9]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45121 entries, 0 to 45120
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   question  45121 non-null  object
 1   answer    45121 non-null  object
dtypes: object(2)
memory usage: 705.1+ KB


In [10]:
train_df.isnull().sum()

question    0
answer      0
dtype: int64

In [11]:
train_df.describe()

question answer
count                                               45121  45121
unique                                               3148  43882
top     What is the most horrible thing you have seen ...   No\n
freq                                                   91     67

In [12]:
import pandas as pd

In [13]:
duplicated_questions = train_df['question'].value_counts().nlargest(5).index
duplicated_questions_df = train_df[train_df['question'].isin(duplicated_questions)]
duplicated_questions_df[['question', 'answer']]

question  \
24              How much will liposuction cost in India?   
237    What are the ways the "all you can eat" rules ...   
301    What is the most horrible thing you have seen ...   
338             How much will liposuction cost in India?   
515    What are the ways the "all you can eat" rules ...   
...                                                  ...   
44351         What are some less known facts about Mars?   
44515  How and from where can I buy cheap flight tick...   
44697  What are the ways the "all you can eat" rules ...   
44723         What are some less known facts about Mars?   
44800         What are some less known facts about Mars?   

                                                  answer  
24     The cost of liposuction in India can vary depe...  
237    My first experience with buffet places were ca...  
301    [LINKED_TEXT: Anxiety ] [URL: https://lifeideo...  
338    Liposuction Surgery Cost in India - Alvura Cos...  
515    Congratulations — this wins the “second-sillie...  
...                                                  ...  
44351  13 Interesting Facts About Mars That Will Amaz...  
44515  WayAway is a renowned online travel agency tha...  
44697  The best one I saw, when I was working at a bu...  
44723  It is the only planet, of which we know, that ...  
44800  What are some mind-blowing facts about the pla...  

[448 rows x 2 columns]

# Data Cleaning

Spelling errors - `Accuracy`


Irrrelavant HTML tags - `Validity`

Emojis - `Validity`

Presence of \n  - `Consistency`



> Data Cleaning Order
1. Quality -> Completeness
2. Tidiness
3. Quality -> Validity
4. Quality -> Accuracy
5. Quality -> Inconsistency



In [15]:
train_df1 = train_df.copy()

In [16]:
import re

In [17]:
def remove_html_tags(text):
    return re.sub(r'<.*?>', '', text)

# Apply HTML tag removal
train_df1['question'] = train_df1['question'].apply(remove_html_tags)
train_df1['answer'] = train_df1['answer'].apply(remove_html_tags)

In [18]:
# Compile the emoji pattern
EMOJI_PATTERN = re.compile(
    "(["
    "\U0001F1E0-\U0001F1FF"  # flags (iOS)
    "\U0001F300-\U0001F5FF"  # symbols & pictographs
    "\U0001F600-\U0001F64F"  # emoticons
    "\U0001F680-\U0001F6FF"  # transport & map symbols
    "\U0001F700-\U0001F77F"  # alchemical symbols
    "\U0001F780-\U0001F7FF"  # Geometric Shapes Extended
    "\U0001F800-\U0001F8FF"  # Supplemental Arrows-C
    "\U0001F900-\U0001F9FF"  # Supplemental Symbols and Pictographs
    "\U0001FA00-\U0001FA6F"  # Chess Symbols
    "\U0001FA70-\U0001FAFF"  # Symbols and Pictographs Extended-A
    "\U00002702-\U000027B0"  # Dingbats
    "])"
)

def remove_emojis(text):
    return EMOJI_PATTERN.sub(r'', text)

    # Apply emoji removal to the 'question' and 'answer' columns
train_df1['question'] = train_df1['question'].apply(remove_emojis)
train_df1['answer'] = train_df1['answer'].apply(remove_emojis)

In [19]:
def remove_newlines(text):
    return text.replace('\n', ' ')

# Apply the function to your dataset
train_df1['question'] = train_df1['question'].apply(remove_newlines)
train_df1['answer'] = train_df1['answer'].apply(remove_newlines)

In [22]:
# Save the cleaned DataFrame to a CSV file
train_df1.to_csv('/content/cleaned_train_df.csv', index=False)

# Load the cleaned DataFrame from the CSV file
cleaned_df = pd.read_csv('cleaned_train_df.csv')

# Display the first few rows
print(cleaned_df.head())

                                            question  \
0  How can a piece of paper folded 103 times be l...   
1  Did a landlord ever offer to lower your rent s...   
2  If the Von Karman line is 62 miles above, then...   
3  Do soldiers pick up fallen enemy weapons to us...   
4          How do I get my son to stop making music?   

                                              answer  
0  It isn’t.  It just has a stack height longer t...  
1  Did a landlord ever offer to lower your rent s...  
2  No. The Karman line has nothing to do with the...  
3  It’s against the rules of war so no, if they r...  
4  Wha- if he loves making music, let him make mu...  
